File: src/ros2_bridge/mappers/foxglove_msgs/voxel_grid.rs
//! Typed mapper for `foxglove_msgs/msg/VoxelGrid`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn voxel_grid_to_bus(msg: ros_env::foxglove_msgs::msg::VoxelGrid) -> crate::foxglove_msgs::msg::v1::VoxelGrid {
    crate::foxglove_msgs::msg::v1::VoxelGrid {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.frame_id),
        pose: Some(crate::ros2_bridge::mappers::foxglove_msgs::pose::pose_to_bus(msg.pose)),
        row_count: msg.row_count,
        column_count: msg.column_count,
        cell_size: Some(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_bus(msg.cell_size)),
        slice_stride: msg.slice_stride,
        row_stride: msg.row_stride,
        cell_stride: msg.cell_stride,
        fields: msg.fields.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::packed_element_field::packed_element_field_to_bus).collect(),
        data: crate::ros2_bridge::mappers::convert::IntoU8Vec::into_u8_vec(msg.data),
    }
}

pub(crate) fn voxel_grid_to_ros(bus: crate::foxglove_msgs::msg::v1::VoxelGrid) -> ros_env::foxglove_msgs::msg::VoxelGrid {
    ros_env::foxglove_msgs::msg::VoxelGrid {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.frame_id),
        pose: crate::ros2_bridge::mappers::foxglove_msgs::pose::pose_to_ros(bus.pose.unwrap_or_default()),
        row_count: bus.row_count,
        column_count: bus.column_count,
        cell_size: crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_ros(bus.cell_size.unwrap_or_default()),
        slice_stride: bus.slice_stride,
        row_stride: bus.row_stride,
        cell_stride: bus.cell_stride,
        fields: bus.fields.into_iter().map(crate::ros2_bridge::mappers::foxglove_msgs::packed_element_field::packed_element_field_to_ros).collect(),
        data: crate::ros2_bridge::mappers::convert::FromByteSeq::from_byte_seq(bus.data),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsVoxelGridMapper;

impl TypedTopicMapper for FoxgloveMsgsVoxelGridMapper {
    type Ros = ros_env::foxglove_msgs::msg::VoxelGrid;
    type Bus = crate::foxglove_msgs::msg::v1::VoxelGrid;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(voxel_grid_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(voxel_grid_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/voxel_grid.py
"""Generated mapper for `foxglove_msgs/msg/VoxelGrid`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.pose import pose_to_bus, pose_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.vector3 import vector3_to_bus, vector3_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.packed_element_field import packed_element_field_to_bus, packed_element_field_to_ros

def voxel_grid_to_bus(msg):
    from robot_bus.foxglove_msgs.msg.v1 import VoxelGrid as BusMsg

    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.frame_id = str(msg.frame_id)
    bus.pose.CopyFrom(pose_to_bus(msg.pose))
    bus.row_count = msg.row_count
    bus.column_count = msg.column_count
    bus.cell_size.CopyFrom(vector3_to_bus(msg.cell_size))
    bus.slice_stride = msg.slice_stride
    bus.row_stride = msg.row_stride
    bus.cell_stride = msg.cell_stride
    bus.fields.extend([packed_element_field_to_bus(x) for x in msg.fields])
    bus.data = bytes(msg.data)
    return bus


def voxel_grid_to_ros(bus):
    from foxglove_msgs.msg import VoxelGrid as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.frame_id = str(bus.frame_id)
    out.pose = pose_to_ros(bus.pose)
    out.row_count = bus.row_count
    out.column_count = bus.column_count
    out.cell_size = vector3_to_ros(bus.cell_size)
    out.slice_stride = bus.slice_stride
    out.row_stride = bus.row_stride
    out.cell_stride = bus.cell_stride
    out.fields = [packed_element_field_to_ros(x) for x in bus.fields]
    out.data = bytes(bus.data)
    return out


class FoxgloveMsgsVoxelGridMapper:
    def ros_msg_type(self):
        from foxglove_msgs.msg import VoxelGrid as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return voxel_grid_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.foxglove_msgs.msg.v1 import VoxelGrid as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return voxel_grid_to_ros(bus)
